joomla/joomla-cms
# [4.0] Add CssIdentifier Rule to future harden the CMS against XSS

### Summary of Changes

This PR adds an additional validation rule for CSS classes to all places where we use them. This has been found in an internal audit and has been reported to the JSST in the past. And we decided now to move it to the public tracker and because you need to have quite high permissions to set it up and we think this should be tested more widely.

Original reporter: @SniperSister, @yvesh and @bembelimen from the JSST and Khoa Bùi Đức Anh

### Testing Instructions

- apply this patch
- go to a module or com_fields where you can add custom css classes.
- add whatever you normally add in there.
- make sure the filter passes that

### Expected result

The filter passes only valid class names

### Actual result

There is no filter and all you put in there could result into an XSS issue in the end.

### Documentation Changes Required

That new rule needs to be documented.

### to-do

- [x] Allow `-1st` `-2st` for the first value of that field
- [x] ~~Allow `--whatever` for the first value of that field~~ (there is no suffix support in 4.x anymore) 
- [x] Make sure the BEM CSS (http://getbem.com/naming/) structure is still supported.
- [x] allow `sm:justify-center`
- [x] add an extra check for value ` 0` as that is a not allowed a 0 without an leading space is allowed.
